zeppelin-web/src/app/notebook/notebook.controller.js

```diff
@@ -1069,8 +1064,55 @@ function NotebookCtrl ($scope, $route, $routeParams, $location, $rootScope,
     $scope.note.config.personalizedMode = isPersonalized
   })
 
+  $scope.$on('$routeChangeStart', function (event, next, current) {
+    if ($scope.note && $scope.note.paragraphs) {
+      _.forEach($scope.note.paragraphs, function (par) {
+        if ($scope.allowLeave === true) {
+          return
+        }
+        let thisScope = angular.element(
+          '#' + par.id + '_paragraphColumn_main').scope()
+
+        if (thisScope.dirtyText === undefined ||
+          thisScope.originalText === undefined ||
+          thisScope.dirtyText === thisScope.originalText) {
+          return true
+        } else {
+          event.preventDefault()
+
+          BootstrapDialog.show({
+            closable: false,
+            closeByBackdrop: false,
+            closeByKeyboard: false,
+            title: 'Do you want to leave this site?',
+            message: 'Changes that you have made will not be saved.',
+            buttons: [{
+              label: 'Stay',
+              action: function (dialog) {
+                dialog.close()
+              }
+            }, {
+              label: 'Leave',
+              action: function (dialog) {
+                dialog.close()
+                angular.element(window).off('beforeunload')
```

Comment:
If we remove event `beforeunload` when page is changed, it can be located in `$destory` as we did befoe?